matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m

```diff
@@ -0,0 +1,174 @@
+function []=fastSLAM_func(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize)
+totalFeatures=MapSize;R=MeasurementNoise;Q=MotionNoise;
+
+
+[timeInterval,finalTime,noFeatures,map,pose,totalParticles,muFeat,muFeatPred,covFeat,covFeatPred,newfeature,w,u,timeStamp,rmax,thmax]...
+            =initializeFastSLAM(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize);
+%% Fast SLAM 
+for t=2:finalTime/timeInterval
+    timeStamp(t)=timeStamp(t-1)+timeInterval;
+    pose(:,t)=motionUpdate_fs(pose(:,t-1),u,Q,timeInterval);
+    meas_ind=featureSearch_fs(noFeatures,map,pose(:,t),rmax,thmax); 
+    y=[];
+    for feature_index = meas_ind       
+        y_new=measurementUpdate_fs(map(:,feature_index),pose(:,t),R);
+        y=[y,y_new];     
+    end
+
+    for particle=1:totalParticles
+       particleSetP(:,particle)=motionUpdate_fs(particleSet(:,particle),u,Q,timeInterval);
+       clear yw hmuw Qw;
+       for j=1:length(meas_ind)
+           i = meas_ind(j);
+           [muFeatPred(:,i,particle),covFeatPred(:,:,i,particle),Hmu]=...
+               ekf_FS1(particleSetP(:,particle),y(:,j),muFeat(:,i,particle),...
+               covFeat(:,:,i,particle),R,newfeature(i));
+           yw(2*(j-1)+1:2*j) = y(:,j);
+           hmuw(2*(j-1)+1:2*j) = Hmu;
+           Qw(2*(j-1)+1:2*j,2*(j-1)+1:2*j) = 2*covFeatPred(:,:,i,particle);
+       end
+       if (exist('Qw'))
+            w(particle) = max(0.000001,mvnpdf(yw,hmuw,Qw));
+       end
+    end
+        newfeature(meas_ind) = 0; 
+        [muParticle,particleSet,muFeat,covFeat]=raoBlackwellFilter_fs(totalParticles...
+            ,w,particleSetP,muFeatPred,covFeatPred);
+        mu_S(:,t) = muParticle;
+    
+     plot_fs(pose,map,y,muFeat,particleSet,t,meas_ind,totalParticles,noFeatures,newfeature);
+       
+end
+
+
+
+    %% Initialization Function   
+    function[timeInterval,finalTime,noFeatures,map,pose,totalParticles,muFeat,muFeatPred,covFeat,covFeatPred,newfeature,w,u,timeStamp,rmax,thmax]...
```

Comment:
I don't often see multiple functions in the same .m file but I've read that it is OK as long as the other functions are sub-functions of the first function. This means the only place the sub-functions are ever called is within the first function. This looks like the case here but you may want to break a few of these out into separate files for use by other examples. For example the "raoBlackwellFilter_fs" and the "featureSearch_fs" functions look like some good generic functions that could be reused in other places.